lib/charms/vault_k8s/v0/vault_kv.py

```diff
@@ -0,0 +1,682 @@
+            egress_subnets.append(str(self.model.get_binding(relation).network.interfaces[0].subnet))
+            relation = self.model.get_relation(relation_name="vault-kv")
+            self.interface.request_credentials(relation, egress_subnets, self.get_nonce())
+
+    def get_nonce(self):
+        secret = self.model.get_secret(label=NONCE_SECRET_LABEL)
+        nonce = secret.get_content(refresh=True)["nonce"]
+        return nonce
+
+
+if __name__ == "__main__":
+    main(ExampleRequirerCharm)
+```
+
+You can integrate both charms by running:
+
+```bash
+juju integrate <vault provider charm> <vault requirer charm>
+```
+"""
+
+import json
+import logging
+from collections.abc import Iterable, Mapping
+from dataclasses import dataclass
+from typing import Any, Dict, List
+
+import ops
+from interface_tester.schema_base import DataBagSchema  # type: ignore[import-untyped]
+from pydantic import BaseModel, Field, Json, ValidationError
+
+# The unique Charmhub library identifier, never change it
+LIBID = "591d6d2fb6a54853b4bb53ef16ef603a"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 11
+
+PYDEPS = ["pydantic", "pytest-interface-tester"]
+
+
+class LogAdapter(logging.LoggerAdapter):
+    """Adapter for the logger to prepend a prefix to all log lines."""
+
+    prefix = "vault_kv"
+
+    def process(self, msg, kwargs):
+        """Decides the format for the prepended text."""
+        return f"[{self.prefix}] {msg}", kwargs
+
+
+logger = LogAdapter(logging.getLogger(__name__), {})
+
+
+class VaultKvProviderSchema(BaseModel):
+    """Provider side of the vault-kv interface."""
+
+    vault_url: str = Field(description="The URL of the Vault server to connect to.")
+    mount: str = Field(
+        description=(
+            "The KV mount available for the requirer application, "
+            "respecting the pattern 'charm-<requirer app>-<user provided suffix>'."
+        )
+    )
+    ca_certificate: str = Field(
+        description="The CA certificate to use when validating the Vault server's certificate."
+    )
+    egress_subnet: str = Field(description="The CIDRs allowed by the role separated by commas.")
+    credentials: Json[Mapping[str, str]] = Field(
+        description=(
+            "Mapping of unit name and credentials for that unit."
+            " Credentials are a juju secret containing a 'role-id' and a 'role-secret-id'."
+        )
+    )
+
+
+class AppVaultKvRequirerSchema(BaseModel):
+    """App schema of the requirer side of the vault-kv interface."""
+
+    mount_suffix: str = Field(
+        description="Suffix to append to the mount name to get the KV mount."
+    )
+
+
+class UnitVaultKvRequirerSchema(BaseModel):
+    """Unit schema of the requirer side of the vault-kv interface."""
+
+    egress_subnet: str = Field(
```

Comment:
Same here

Reply:
I created a tickets to use pydantic validation